File: lib/obstacle.js
const Vector = require('./vector');
const Collider = require('./collider');

class Obstacle extends Collider {
  constructor(pointArray){
    super(pointArray);
  }

  handleCollison(targetpoint, position){
    const collidingEdge = this.getEnteringEdge(targetpoint, position);
    let outsidepoint = this.getClosestOutsidePoint(collidingEdge, targetpoint);
    let movementDirection = outsidepoint.difference(position);
    targetpoint = position.sum(movementDirection);
    return targetpoint;
  }
}
module.exports = Obstacle;

File: Firmware/ESP32/obstacle.cpp
#include "obstacle.hpp"

Obstacle::Obstacle(std::vector<Vector2D> points) : Collider(points) { }

Vector2D Obstacle::handleCollision(Vector2D targetPoint, Vector2D position)
{
    Edge collidingEdge;
    // ignore return value - we already know it's colliding
    getEnteringEdge(targetPoint, position, &collidingEdge);
    return getClosestOutsidePoint(collidingEdge, targetPoint);
}